### ISSUE TITLE
A question about decomposed_coordinates_and_features

### ISSUE THREAD

**Issue by somebodyus** [NVIDIA/MinkowskiEngine#134]
Suppose we have:

```
>>> coords, feats = ME.utils.sparse_collate([coords_batch0, coords_batch1], [feats_batch0, feats_batch1])
>>> A = ME.SparseTensor(feats=feats, coords=coords) 
>>> list_of_coords, list_of_featurs = A.decomposed_coordinates_and_features
```

Should `torch(list_of_featurs, dim=0) == feats` ? I think it should be true, but I verified that they are not equal. Could anyone tell me why?

**Comment by chrischoy** [comment]
There are kind of serveral reasons

On a dense tensor, random access by a coordinate is straight forward (ravel), but on a sparse tensor, you could not use such memory mapping. Instead, we use a hash table key: coordinate -> value: row index.

When we decompose a sparse tensor by batch index, internally we use a global pooling mapping that maps everything to the origin. This process requires going over all items in a hash map and finding the origin for each batch. This order, however, is pseudo-random as it is defined by the hasy key.

Once we find the global pooling mapping, we use that to extract features and coordinates, and thus the order is somewhat random.

However, the feature at the corresponding coordinate never changes. If (coordinate, feature) pairs are unique, then a sparse tensor is equivalent to other sparse tensors with different order of (coordinate, feature) pairs.

**Comment by somebodyus** [comment]
Thank you very much for your reply. I am doing a point-wise classification task. If using this method give a random-order list, it cannot be compared with the torch tensor ground truth. So I find in your sample code, you use something like `torch.CrossEntropy(output.F, label)`. 

I am wondering is it OK to warp the ground truth in a sparse tensor. And doing the batch-wise comparison like this:

```
slabel = ME.SparseTensor(feats=data_dict['label'], coords=data_dict['coords']).to(args.gpu) 
torch.CrossEntropy(output.features_at(batch_id), label.features_at(batch_id))
```

The reason for doing this is that I need to do batch-wise comparison and the `.features_at` method can easily give the n-th data. If this operation is not feasible, I need to first find the row indices and mask out the torch tensor.

**Comment by chrischoy** [comment]
Ah that's pretty clever.

Yes, you can do that.

To speed up and reuse the hash map generated within the output. Try

```
slabel = ME.SparseTensor(feats=labels, coords_key=output.coords_key, coords_man=output.coords_man)
torch.nn.CrossEntropy(output.features_at(batch_id), slabel.features_at(batch_id))
```
